File: LeetCode/Python/69.sqrt-x.py
# @before-stub-for-debug-begin
from python3problem69 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=69 lang=python3
#
# [69] Sqrt(x)
#

# @lc code=start


class Solution:
    def mySqrt(self, x: int) -> int:
        if x <= 1:
            return x
        
        # Since the Square root can't be half of a number
        l, r = 0, x // 2

        m = 0
        while l <= r:
            # (r + l) will cause overflow in other languages
            m = l + (r - l) // 2

            # 2 * 2 <= 5 < 3 * 3
            if m * m <= x < (m + 1) ** 2:
                return m
            # 3 * 3 > 5
            elif m * m > x:
                r = m - 1
            # 1 * 1 < 5
            else:
                l = m + 1
# @lc code=end

File: LeetCode/CPP/69.sqrt-x.cpp
/*
 * @lc app=leetcode id=69 lang=cpp
 *
 * [69] Sqrt(x)
 */

// @lc code=start
class Solution
{
public:
    int mySqrt(int x)
    {
        if (x <= 1)
            return x;
        long long l = 0, r = x / 2, m, ans = -1;

        while (l <= r)
        {
            m = l + (r - l) / 2;
            if (m * m < x)
            {
                ans = m;
                l = m + 1;
            }
            else if (m * m > x)
            {
                r = m - 1;
            }
            else
                return m;
        }
        return ans;
    }
};
// @lc code=end
